### ISSUE TITLE
🚀 [Feature Request] Run/ Walk mobs can run or walk during random movement

### ISSUE THREAD

**Issue by al3xc1985** [cmangos/issues#3699]
### Feature Details

Run/ Walk mobs can run or walk during random movement
This is extremly usefull cause a lot of them are behaving this way The system choose at a based chance to run or walk
From experience I can say it's more like 70% walk 30% run chance.
Many Many mobs in azeroth can run/walk. I have videos with most of the azeroth and I captured most of the mobs until now that do run randomly

Best example are children in goldshire  in the back side of the blacksmith

Video can be provided if needed
https://youtu.be/WU-g7rvQUbc

The bear and the owl have those moments of running, and u can see the we have a patrol that is not modified by netowrk packets syncronization.

### Visualizations

1

### Benefits and Purpose

1

### Core Versions

- [X] 1.1.12 (Classic)
- [X] 2.4.3 (The Burning Crusade)
- [X] 3.3.5a (Wrath of the Lich King)

**Comment by Grz3s** [comment]
`random movement: running` is a old known missing feature that ive been asking for several years.
- it may happen randomly walk/run
- may be constant (only running) 
- may also be part of scripted event: (and golshire kids are one of them) 
they do waypoints with 2 stops:
- 1st with randomovement walking on timer
- 2nd with randomovement running on timer - and this part is not implemented ...
Reason for . .why its not added yet .. is (as far as ive been told).. our legit data does not show anywhere change between walk/run they do - or simply noone is brave enough to touch randommovement itself. (it doesnt work well anyway).

**Comment by Warlockbugs** [comment]
So far all observations on this subject have been inconsistent and most likely this is the result of some timing mismatch between the server and client with client trying to catching up with the server, which visually speeds up the unit (same thing happens during network lag). We have not had a single packet for analysis and implementation yet.

**Comment by Grz3s** [comment]
OK - for kid https://www.wowhead.com/wotlk/npc=797/bo#comments in goldshire we could explain it by sain, that it hapenns only because He runs to 1st stop point .. and for some reason - running flag is not removed (will do randommovement running). However: How to explain that after he runs back to 2nd stoping point - he will miracly change to random walk ..
Situation repeats .. and never changed for smth else.

Other guy i cant find any explanation:
https://www.wowhead.com/wotlk/npc=29124/lifeblood-elemental
this one constantly do random running... 
https://youtu.be/jf8h8YZJsq4?t=41 -- top right guy for example.
elementals are good koz they have diff animation for walking (hands down) .. running (hands lifted)
Theyre many more examples:
```
28747 - all with aura 52385 only!!!! (movementtemplateid = 0)
28851 - all (movementtemplateid = 165)
28861 - only with randommovemnt
29124 - only with randommovemnt
```
I know that subject is difficult... but cannot be denied.

Side note:
Other diff types of randomovement exists, that not inidicating anything in packets... and cannot be tracked without recording and observing creatures...
1 example is 
![4234234](https://github.com/cmangos/issues/assets/2374069/be31b736-a3c4-4de4-9284-90658a92c7c2)
Creature will have big walk distance on X and short on Y. They will walk anywhere inside that border line but never cross it... ( i can also give  many examples for this one .. tauren starting zone guard ,, as easiest to see.)

**Comment by al3xc1985** [comment]
@Warlockbugs I found so many mobs/npc's that walk to a certain point and run from that point to next point,
and if u say that it's a network thing, let me tell u that u are wrong here!
Why?
Because I stood and watched the mob doing his path more than 3 times, and every single time those parts of his run are in there, followed bu walking.

This is not concidence anymore

And dunno on tbc time, but on wotlk time, we have that a lot

**Comment by Grz3s** [comment]
running from point A to B is not a pronlem.. and we can do that.
Issue only exist for randommovement that no mater what ill do.. always change to walk flag. (which is not correct for some)

**Comment by al3xc1985** [comment]
in first place that is the problem with random walk/run., there are also a lot of mobs that I studied and they do have this behaviour , and choose between them.. I believe it's more like 60% walk chance 40% run chance from what I have seen.

But dunno why I understood the thing with the run into scripts..I may understood it wronge, but this is not the main thing on the subject

**Comment by al3xc1985** [comment]
@Grz3s about the random movement I have specified in a different thread that mobs can choose to walk in line, like they were having path, but it's just a random line movement. not just walking around in a given radian circle

**Comment by AnonXS** [comment]
Can we get some videos showcasing this behavior?

**Comment by al3xc1985** [comment]
I am recording BRD right now, and it is very laborious stuff, a lot of emotes and patrols, but I shall think of it, an come back with at least a couple of them for the subject of this thread

**Comment by al3xc1985** [comment]
if @Grz3s some some examples faster I beg him to lay them down

**Comment by Tobschinski** [comment]
> Side note: Other diff types of randomovement exists, that not inidicating anything in packets... and cannot be tracked without recording and observing creatures... 1 example is 
![4234234](https://private-user-images.githubusercontent.com/2374069/321797724-be31b736-a3c4-4de4-9284-90658a92c7c2.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI5MDIyNjEsIm5iZiI6MTcxMjkwMTk2MSwicGF0aCI6Ii8yMzc0MDY5LzMyMTc5NzcyNC1iZTMxYjczNi1hM2M0LTRkZTQtOTI4NC05MDY1OGE5MmM3YzIuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MTJUMDYwNjAxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjMyNzBjNTkxZTI4YWJkZTVjYWRjMTRkYjI4NzUwYWJiM2UyMWIzODVmODc5Njg3YjFkZDRmNTI0MDBiZWQzMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.aj57aIo0A3vxhk4vi4V-MMydtVorO5LZAaghbkoJtPM) 
Creature will have big walk distance on X and short on Y. They will walk anywhere inside that border line but never cross it... ( i can also give many examples for this one .. tauren starting zone guard ,, as easiest to see.)

![image](https://github.com/cmangos/issues/assets/4753528/2ed01c41-ef68-48e6-9735-c52b671f6f0c)

Do you think it could simply be a script that does random movement around 2 points, instead of 1? I've seen this behavior in a lot of cats and this was the first explanation I could come up with, without checking any further..

**Comment by Warlockbugs** [comment]
@al3xc1985
My post is strictly about rare occurrences of wandering unit running instead of walking to the next wander point. What happens next is that unit resumes to walking again to next point of wander after that. So far any observations on that amounted simply to a noise and not single packet with a proof or details presented for implementation, so i grew increasingly skeptical over the years about this "feature". I'll believe it when i see it.
- Client has a tendency to pick displayed movement animation by itself for a unit model based on its speed. You can observe this via using .modify speed on yourself. For your regular running speed, at speed modifier value over approx. 1.5 your animation switches to "sprint" kind. There may be similar client-side mechanic for units with unusually high walking speed, that can amount to the observed effect
- Network hiccups, client visually speeds up all the movement that lags behind the server
A combination or either of these two can result in observed "random runs during wandering" that we are struggling to catch in the traffic.

**Comment by Warlockbugs** [comment]
@Grz3s 
The best lead we can possibly have is movement speed values and changes, as that can possibly affect what kind of animation client picks for display.

**Comment by Grz3s** [comment]
@Warlockbugs 
Nah. I took elemental for reason here ;) ... Changing his walk speed wont change his animation (hands will be still down)...
Also checking them in sniff ... we see his walk and run speed .. that shows default values..
@Tobschinski 
simply  - no.
You would have 2 repeating points from time to time... also you would have 2 circles (like on your pic..).. 
From points i gathered it shows, that they walk to any pos inside that border.
![321797724-be31b736-a3c4-4de4-9284-90658a92c7c2](https://github.com/cmangos/issues/assets/2374069/a741ff6d-9e29-44bd-8391-ea446844ad1e)

**Comment by Warlockbugs** [comment]
@Grz3s 
Alright. I do suspect that answer to this is purely technical and related to how client chooses to process movement data. Maybe there is some special instance of "finish this movement asap and sync position" similar to regular network lag recovery mechanism.

**Comment by al3xc1985** [comment]
the thing is that if u observer the kids from back of blacksmith for example, when they are in group one or two are walking, one is running, one is standing, and this cycle goes random.
Is that possible @Warlockbugs with that network compensation to see them doing walking and running in same time?

**Comment by Warlockbugs** [comment]
The children in goldshire is a special case of formation with random wandering on top of it. When they are not wandering they should be trying to catch up with the leader of formation with a certain offset. We do not support movement generator stacking and mutation (yet).

**Comment by al3xc1985** [comment]
@Warlockbugs & @AnonXS I uploaded a video with 2 mobs in same place doing it one after the other one.
And I need to add that my internet is 1000 Mb/s from my privider and tested on speed test with 970+ acitve Mb/s. If that helps

Also after this video I stood and see if they do it again and yes, I catched the owl one more time, but no more, time to work on recordings

**Comment by al3xc1985** [comment]
I can add more in different zones if needed, but I find those 2 when I was recording the whole felwood zone

**Comment by al3xc1985** [comment]
I also need to add the I found mobs that just run when I get close to them, as a compensation for network syncronization, but in my vid is not the case

**Comment by al3xc1985** [comment]
I can see mobs randomly funning even in classic. But the chance that a mob can random run during his random movement is like 20% running mod to 80% normal walking in my opinion
@killerwife I ask for your opinion in this please

**Comment by al3xc1985** [comment]
More to say about is that I found that they choose to random run or walk to short distance and to long distance alike. So it is not a distance based algorithm.
But I can say that walk is a higher chance than run.

**Comment by al3xc1985** [comment]
It's preaty hard to guess the percent of witch they choose to run or walk, but after a lot of recording, I can say as I said before 20% chance to run, 80% chance to walk.

more than 80 of mobs that do random movement, have this "flag" or "feature"added

**Comment by miraco** [comment]
I*ve got another NPC that seems to change between running and walking on random Movement.
Maxx A. Million Mk. II
Also his waypoint movement seems to be like Grz3s talked about in here.

![Image](https://github.com/user-attachments/assets/27c949c6-5780-42b5-8361-46fdf3bc1bae)

The one waypoint i cut off on the left is related to a RP between the robot and Bot-Specialist Alley, where the bot runs to far away and Alley is calling him back.

Video of the random Movement:
https://youtu.be/7Q4nVNB3KUg 

I can also provide sniffs if needed. after watching him for around 20 minutes i can say it looks like its doing more Running then Walking.

overall i feel like random movement still is a big mystery with many different cases.

**Comment by al3xc1985** [comment]
As I see the problem is:
1. Most of the mobs, beside  humanoids, have random run/walk setting by default
2. BUT, some of them need like in your example to have an adjustement setting, where u can force the random movement system to use percentages of happening. I mean 60% run set, would result that 40% of the time will walk, it will random run 60% of the time.

Even thou it doesn't appear in sniffs, I am sure that through the same packets or how ever is called, bl!zz is doing all this tricks

**Comment by Grz3s** [comment]
It took some time .. but we finally got tools to to it... 
With `CREATURE_EXTRA_FLAG_RUN_DURING_WANDER` - its 15% run
For more % we can use db scripts `20  - Set RandomMovement + RUN` 

Now question would be .. where we want to add it..
all Bears, Boars, Wolfes.. ? (these i know do that) .. or should we be careful & go 1 by 1

**Comment by killerwife** [comment]
Please only add it where its confirmed, because this topic will need more research and i want a solid foundation to build on.

**Comment by miraco** [comment]
> It took some time .. but we finally got tools to to it... With `CREATURE_EXTRA_FLAG_RUN_DURING_WANDER` - its 15% run For more % we can use db scripts `20 - Set RandomMovement + RUN`
> 
> Now question would be .. where we want to add it.. all Bears, Boars, Wolfes.. ? (these i know do that) .. or should we be careful & go 1 by 1

I agree with KW on this, we should only add it to npcs where its defnitly confirmed.

But i have a question about this the npc in netherstorm seems to have 80% run and 20% walk
to get this i have to let the npc start a db script on spawn with command 20 and dataint 1 and dataint2 1 (to get back rnd movement on evade)? 
where can i set the % that the npc keeps switching between walk and run?

**Comment by killerwife** [comment]
You cannot. That extra flag is fixed. For now

**Comment by miraco** [comment]
Okay that is good to know then I have to wait before updating that npc

**Comment by Grz3s** [comment]
test this:
```
-- Maxx A. Million Mk. II 19588
UPDATE creature SET position_x = 3353.966309, position_y = 3735.008545, position_z = 140.986359 WHERE guid = 5306836;
-- Generic for Random run+walk (dist 7)
DELETE FROM dbscript_random_templates WHERE id = 20388;
INSERT INTO dbscript_random_templates (id, type, target_id, chance, comments) VALUES
(20388,1,30690,65,'GENERIC - SET RandomMovement (Run) - Random Script 1'), -- 65%
(20388,1,30691,35,'GENERIC - SET RandomMovement (Walk) - Random Script 2'); -- 35%
DELETE FROM dbscripts_on_relay WHERE id IN (30690,30691);
INSERT INTO dbscripts_on_relay (id, delay, command, datalong, datalong2, datalong3, buddy_entry, search_radius, data_flags, dataint, dataint2, dataint3, dataint4, x, y, z, o, comments) VALUES
(30690,0,20,1,7,0,0,0,0x04,1,0,0,0,0,0,0,0,'GENERIC - SET RandomMovement + RUN'),
(30691,0,20,1,7,0,0,0,0x04,0,0,0,0,0,0,0,0,'GENERIC - SET RandomMovement');

-- EAI
DELETE FROM creature_ai_scripts WHERE creature_id IN (19588);
INSERT INTO `creature_ai_scripts` (`id`,`creature_id`,`event_type`,`event_inverse_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`event_param5`,`event_param6`,`action1_type`,`action1_param1`,`action1_param2`,`action1_param3`,`action2_type`,`action2_param1`,`action2_param2`,`action2_param3`,`action3_type`,`action3_param1`,`action3_param2`,`action3_param3`,`comment`) VALUES
-- Maxx A. Million Mk. II 19588
('1958801','19588','1','0','100','1','2000','7000','6000','11000','0','0','53','-20388','0','0','0','0','0','0','0','0','0','0','Maxx A. Million Mk. II - Random Script on OOC Timer');
UPDATE `creature_template` SET AIName = 'EventAI' WHERE `entry` IN (19588);
```

can change % to whatever u like ... also timer & distance can be ajusted

**Comment by al3xc1985** [comment]
@Grz3s can we close this one? I think we have what it needs now! What do you think?
